src/crush/CrushWrapper.cc

```diff
@@ -1647,12 +1647,22 @@ int CrushWrapper::update_device_class(int id,
                                       const string& name,
                                       ostream *ss)
 {
+  assert(item_exists(id));
+  auto old_class_name = get_item_class(id);
+  stringstream ts;
+  if (old_class_name &&
+      old_class_name != class_name && // so command is idempotent
+      class_is_in_use(get_class_id(old_class_name), &ts)) {
+    *ss << "osd." << id << " has already bound to class '" << old_class_name
+        << "', which is currently " << ts.str();
+    return -EBUSY;
```

Comment:
This works as a guard to  prevents the problem below from happening:

```
ceph osd crush set-device-class hdd osd.0
ceph osd crush rule create-replicated foo-hdd1 default host hdd
ceph osd crush set-device-class ssd osd.0
```

Before this change, the above ops will be considered as legal and we'll silently change the class of osd.0 from **hdd** into **ssd**.
But actually it is not unsafe to do so as the old class of osd.0 might be  in use by crush rule **foo-hdd1**.

>  If the class is wrong, what is the user supposed to do?


In case user want to (re)use "set-device-class" to reset the class of specific osds, make sure the old class is not currently in use by any crush rule. Otherwise we fail and user shall delete the crush rule that is references the old device class first before he can proceed.